### ISSUE TITLE
ActiveSupport 4.2 - `JSON.dump` broken for Structs

### ISSUE THREAD

**Issue by arnvald** [rails/rails#20775]
Hi everyone,

Today I've noticed that ActiveSupport 4.2 breaks `JSON.dump` ("stack level too deep") for the arrays/hashes containing structs.

Example:
```ruby
irb(main):001:0> require 'json'
=> true
irb(main):002:0> obj = Struct.new(:key).new
=> #<struct key=nil>
irb(main):003:0> JSON.dump([obj])
=> "[\"#<struct key=nil>\"]"
irb(main):004:0> require 'active_support/json'
=> true
irb(main):005:0> JSON.dump([obj])
SystemStackError: stack level too deep
```

Stack trace shows that this line `activesupport-4.2.0/lib/active_support/core_ext/object/json.rb:34:in `to_json_with_active_support_encoder'` is called over and over.

I tried to figure out why is it failing, but I couldn't find the root of the problem.
If you have any hint on how I can provide more useful information, let me know.

This issue appears in ActiveSupport 4.2.x (I've checked 4.2.0 and 4.2.3), but it does not appear in 4.1.x

**Comment by pascalbetz** [comment]
Here is a runnable test that shows the aforementioned StackOverflow. If ``` ActiveSupport::JSON``` is loaded then the test fails. If the line is removed then the test passes.
Only happens for Struct (at least not for the few basic classes if tested).

```ruby
begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'
  gem 'activesupport'
  gem 'minitest'
end

require 'json'
require 'active_support'
require 'minitest/autorun'

class BugTest < Minitest::Test

  def test_stack_error
    # Enable/Disable to see difference
    ActiveSupport::JSON
    obj = Struct.new(:key).new
    assert_equal "\"#<struct key=nil>\"", JSON.dump(obj)
  end

  def _test_other_objects
    ActiveSupport::JSON
    [Object.new, "", false, 1, Time.now, Date.today].each do |obj|
      JSON.dump(obj)
    end
  end
end

```

**Comment by fenec** [comment]
This issue happens because `Enumerable` to the list of classes that are provided with `ActiveSupport` `to_json_with_active_support_encoder` method. Which is kinda weird. Anyway, I'll try to dig into this thing and maybe find a solution

**Comment by prathamesh-sonpatki** [comment]
This issue is fixed on master in - https://github.com/rails/rails/commit/d5bddc1b2d3d794b2eddcb7309f41f87a8d665c1.

**Comment by chancancode** [comment]
Ahhh I know what's up. We introduced this bug in #10278. The intention of [this line](https://github.com/rails/rails/blob/4-2-stable/activesupport/lib/active_support/core_ext/object/json.rb#L29) is to ["compete" with these JSON gem additions](https://github.com/flori/json/blob/master/lib/json/pure/generator.rb#L324-L519) so we get to do our thing before JSON gem does.

However, since JSON gem doesn't actually define `#to_json` on `Enumerable`, the `alias_method_chain` ended up aliasing `Object#to_json` (which is in turns an alias of `Object#to_json_with_active_support_encoder`) as `Enumerable#to_json_without_active_support_encoder`, so when we later try to call that method, we created an infinite loop. (It is not immediately clear to me *why* this happens, I would expect the AMC on `Enumberable` to fail with a `NameError`, it's still a mystery to me why it ended up grabbing `Object#to_json` there.)

I don't have any great suggestions yet, I'll keep pondering. We should fix this!

cc @codeodor

**Comment by codeodor** [comment]
@prathamesh-sonpatki running the provided test case against latest master still has the issue for me.

Thanks @chancancode for bringing it to my attention. I will see if I can figure out a fix.

**Comment by codeodor** [comment]
My apologies, I ran it against 4.2.3, not master. Trying that now.

**Comment by codeodor** [comment]
Confirm fixed on master. I didn't test enough to confirm whether that commit @prathamesh-sonpatki mentioned its the fix, but I will take their word for it.

**Comment by arnvald** [comment]
@prathamesh-sonpatki is right, I've just checked it in master and it works.

If a new release in 4.2.x branch is planned, is there a way to get this fix there? Or does it need to wait for Rails 5?

**Comment by prathamesh-sonpatki** [comment]
@codeodor @chancancode @arnvald I thought of a naive solution. If Enumerable is not supporting `to_json` in JSON gem then we can handle that case separately because it that case we want to and have to call only `ActiveSupport::JSON#to_json_with_active_support`. Thoughts?

**Comment by chancancode** [comment]
My hunch is that it's subtly wrong; but please give it a try if you have time! :smile: 

My worries are: 

- `JSON` gem [checks if an object defines `to_json`](https://github.com/flori/json/blob/master/ext/json/ext/generator/generator.c#L867) and call it, otherwise it will call `to_s`, so by defining `to_json` on a `BasicObject`we would have changed how `JSON.dump` work already, but I don't 100% know what the implications are, yet (I guess maybe the `if options.is_a?(::JSON::State)` should call `to_s` in that case..)

- It's not exactly the same thing as the `prepend`: if there is a `to_json` somewhere above us, we won't be calling it anymore

That seems "wrong" – the

**Comment by chancancode** [comment]
I also still don't know why the `alias_method_chain` on `Enumerable` isn't failing with an error, if someone figured that out please educate me :grin:

**Comment by codeodor** [comment]
I think it's something to do with the fact that we create this for `Object`, which gets on to `Struct`. But we create on Enumerable too, which also gets on to `Struct`.

For instance, if I ensure it only makes it on to the object one time, then the test case here works. The trouble with this approach though, is that it breaks a dozen tests in Rails.

Also, the behavior is not the same for `Array` and `Hash` (which I would expect) ... but those seem to work fine.

**Comment by codeodor** [comment]
@prathamesh-sonpatki  I *thought* I tried that in a convoluted way, but when I break it out as you mentioned, it seems OK so far.

**Comment by codeodor** [comment]
Also should point out that with ActiveSupport::JSON not in there, `assert_equal "\"#<struct key=nil>\"", JSON.dump(obj)` is the wrong answer anyway. It's just that at least it doesn't endlessly recurse.

**Comment by codeodor** [comment]
#20791 fixes this

**Comment by chancancode** [comment]
Fixed in 49c613463b758a520a6162e702acc1158fc210ca; tests added in d6142c74d1f9decc2687aa388899a76a03ab0e49.

Thank you so much for reporting this!!! :green_heart: :yellow_heart: :heart: :blue_heart: :purple_heart:
